### PR TITLE
Converted app/javascript/mastodon/utils/ folder to TypeScript

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Converts remaining files in `app/javascript/mastodon/utils/` to TypeScript.

Also removes the old browser polyfills from two places. ⚠️ Beware that this is a behavior change!

* `app/javascript/mastodon/utils/notifications.ts`
* `app/javascript/mastodon/utils/scrollbar.ts`

### CODE REVIEW

File: app/javascript/mastodon/utils/notifications.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Handles browser quirks, based on
+// https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API
+
+interface WritableNotification {
+  permission: NotificationPermission;
+}
```
Comment (JoshuaKGoldberg) [inline app/javascript/mastodon/utils/notifications.ts:6]: Similar issue to https://github.com/mastodon/mastodon/pull/27894/files#r1396032427: TypeScript assumes a newer / more strict browser support range.

File: app/javascript/mastodon/utils/scrollbar.ts [hunk 2]
```diff
@@ -16,20 +12,20 @@ const getActualScrollbarWidth = () => {
   outer.appendChild(inner);
 
   const scrollbarWidth = outer.offsetWidth - inner.offsetWidth;
-  outer.parentNode.removeChild(outer);
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  outer.parentNode!.removeChild(outer);
```
Comment (JoshuaKGoldberg) [inline app/javascript/mastodon/utils/scrollbar.ts:16]: [`Element.remove()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/remove) exists now, but I'm hesitant to change behavior in a TypeScript conversion PR. I've learned the hard way that that's how I tend to introduce bugs 😄

File: app/javascript/mastodon/utils/html.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// NB: This function can still return unsafe HTML
+export const unescapeHTML = (html: string) => {
+  const wrapper = document.createElement('div');
+  wrapper.innerHTML = html
+    .replace(/<br\s*\/?>/g, '\n')
+    .replace(/<\/p><p>/g, '\n\n')
+    .replace(/<[^>]*>/g, '');
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply app/javascript/mastodon/utils/html.ts:7]: This is existing code.

File: app/javascript/mastodon/utils/notifications.ts (hunk 1 above)
Comment (renchap) [reply app/javascript/mastodon/utils/notifications.ts:6]: Is all of this still needed?

We no longer support really old browsers, but I am having trouble figuring out if `Notification.permission` might be unimplemented on some modern browsers

File: app/javascript/mastodon/utils/scrollbar.ts (hunk 2 above)
Comment (renchap) [reply app/javascript/mastodon/utils/scrollbar.ts:16]: Support is very wide, you can use it and it will be cleaner :)

**Review by renchap — COMMENTED** [review]
Thanks a lot for your contributions, I am always happy to see more Typescript.

I will review them (slowly probably).

I left a few comments on this one, it should be mergeable once they are fixed.

Also `mastodon/utils/log_out.ts` seems to not be typed and makes the CI checks fail.

File: app/javascript/mastodon/utils/scrollbar.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [reply app/javascript/mastodon/utils/scrollbar.ts:16]: Awesome thanks 🙂 

Just confirming - for this and other runtime fixes, would you prefer them in this PR or a separate one? I personally prefer separating type system changes from runtime but also understand the extra overhead isn't ideal.

Comment (renchap) [reply app/javascript/mastodon/utils/scrollbar.ts:16]: I am fine with doing it in the same PR, as long as you add a comment pointing those changes

File: app/javascript/mastodon/utils/notifications.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply app/javascript/mastodon/utils/notifications.ts:6]: I haven't been able to find a situation where it isn't.

**Review by renchap — APPROVED** [review]

**Pull request merged**